android/src/main/java/co/airbitz/fastcrypto/MoneroAsyncTask.java

```diff
@@ -67,18 +68,50 @@ protected Void doInBackground(Void... voids) {
                     }
                 }
                 connection.connect();
+
+                if (isStopped.get()) { 
+                    promise.reject("Err", new Exception("Download cancelled by user."));
+                    return null; 
+                }
+
                 String contentLength = connection.getHeaderField("Content-Length");
                 int responseLength = Integer.parseInt(contentLength);
                 try (DataInputStream dataInputStream = new DataInputStream(connection.getInputStream())) {
                     byte[] bytes = new byte[responseLength];
-                    dataInputStream.readFully(bytes);
+                    int bytesRead = 0;
+                    int offset = 0;
+
+                    // Check for cancellation periodically during download
+                    while (bytesRead != -1 && offset < responseLength) { 
+                        if (isStopped.get()) {
+                            promise.reject("Err", new Exception("Download stopped by user."));
+                            return null;
+                        }
+                        bytesRead = dataInputStream.read(bytes, offset, responseLength - offset);
+                        offset += bytesRead;
+                    }
+                    
+                    // Check for cancellation after download is complete
+                    if (isStopped.get()) { 
+                        promise.reject("Err", new Exception("Processing are stopped"));
+                        return null; 
+                    }
+
```

Comment:
This seems to be repeated twice. Would it make sense to put it in a function (e.g. `readFully`) to avoid code repetition?

Reply:
> This seems to be repeated twice.

Do you mean this code is repeated? 

```
if (isStopped.get()) { 
      promise.reject("Err", new Exception("Processing are stopped"));
      return null; 
  }
  ```

Reply:
i think @feri42 means the whole while loop

Reply:
Correct, I mean the whole loop. And there may actually be even more than that at first glance.